tools/extract-formats.py

```diff
@@ -34,42 +50,76 @@ def main(options, args=None, output=sys.stdout, lines=None):
 
         match = typeline.fullmatch(line)
         if match:
-            if message is not None:
+            if match.group('type') == 'tlv':
+                if tlv:
+                    raise ValueError('{}:Found a tlv while I was already in a tlv'
+                                     .format(linenum))
+                tlv = match.group('name')
+                tlv_msg_count = 0
+                continue
+
+            if message and not tlv:
                 raise ValueError('{}:Found a message while I was already in a '
                                  'message'.format(linenum))
+
             message = match.group('name')
+            if tlv is not None and len(match.group('leading')) == 0:
+                tlv = None
             if options.output_types:
-                print("{},{}".format(
-                    match.group('name'),
-                    match.group('value')), file=output)
+                if tlv:
+                    print("{},{},{}".format(
+                        match.group('name'),
+                        match.group('value'),
+                        tlv), file=output)
+                else:
+                    print("{},{}".format(
+                        match.group('name'),
+                        str(match.group('value') or '$')), file=output)
```

Comment:
I think subtypes should just be a single field, ie. no value.  Rather than a double field with a magic '$' sign?